packages/components/src/internal/components/navigation/ProductMenu.tsx

```diff
@@ -13,100 +13,328 @@
-                            currentProductId={model.currentProductId}
-                        />
-                    ))}
-                </div>
-            ));
-        } else {
-            content = model.sections.map(section => (
-                <div key={section.key} className={menuSectionCls}>
-                    <ProductMenuSection
-                        section={section}
-                        config={new MenuSectionConfig()}
-                        currentProductId={model.currentProductId}
-                    />
-                </div>
-            ));
-        }
-    }
+    if (!isLoaded && !hasError) return null;
+    const showFolders = folderItems?.length > 1;
 
     return (
         <DropdownButton
             className="product-menu-button"
             id="product-menu"
             onToggle={toggleMenu}
             open={menuOpen}
-            title="Menu"
+            title={<ProductMenuButtonTitle container={container} folderItems={folderItems} routes={routes} />}
         >
-            <div className={classNames('product-menu-content', { error: !!model?.isError })} onClick={onClick}>
-                <div className="navbar-connector" />
-                {content}
-            </div>
+            {menuOpen && (
+                <ProductMenu
+                    {...props}
+                    className={classNames({ 'with-col-folders': showFolders })}
+                    onClick={onClick}
+                    error={error}
+                    folderItems={folderItems}
+                    showFolderMenu={showFolders}
+                />
+            )}
         </DropdownButton>
     );
 });
+
+export const ProductMenuButton = withRouter<ProductMenuButtonProps>(ProductMenuButtonImpl);
+
+export interface ProductMenuProps extends ProductMenuButtonProps {
+    className: string;
+    error: string;
+    folderItems: FolderMenuItem[];
+    onClick: (evt: MouseEvent<HTMLDivElement>) => void;
+}
+
+export const ProductMenu: FC<ProductMenuProps> = memo(props => {
+    const {
+        className,
+        onClick,
+        error,
+        folderItems,
+        sectionConfigs,
+        showFolderMenu,
+        appProperties = getCurrentAppProperties(),
+    } = props;
+    const { api } = useAppContext<AppContext>();
+    const { container, moduleContext } = useServerContext();
+    const [menuModel, setMenuModel] = useState<ProductMenuModel>(new ProductMenuModel({ containerId: container.id }));
+    const contentRef = useRef<HTMLDivElement>();
+
+    useEffect(() => {
+        if (!menuModel.isLoaded) return;
+        let height = 400; // match navbar.scss product-menu-content min-height
+        const maxHeight = window.innerHeight * 0.8;
+        const sections = Array.from(contentRef.current.getElementsByClassName('menu-section'));
+        sections.forEach(section => (height = Math.max(height, section.clientHeight)));
+        contentRef.current.style.height = Math.min(height, maxHeight) + 'px';
```

Comment:
Might want to add a comment here about the desire display effect.

Reply:
Done